### ISSUE TITLE
1 [B1] Best Practices and Notable Particularities

### ISSUE THREAD

**Issue by esteblock** [soroswap/aggregator#66]
**Description**
Here are some notes on the protocol particularities, comments, and suggestions to improve the
code or the business logic of the protocol in a best-practice sense. They do not in themselves
present issues to the audited protocol but are advised to either be aware of or to be followed
when possible, and may explain minor unexpected behaviors on the deployed project.

- [x] 1.- Aggregator storage function remove_adapter_id is somewhat incorrectly named, to match the other functions it should be remove_protocol_id.

- [x] 2.- Aggregator function calculate_distribution_amounts calls Iterator::sum which
allows overflows for production code. However, since overflow-checks=true is set in the
workspace Cargo.toml , this should not actually create an overflow. Still, it is
recommended that even with this set that code should be defensive of overflow;

- [x] 3.- Specific functions are repeated throughout the aggregator repository, such as check_nonnegative_amount , ensure_deadline , and check_initialized . Having
multiple instances of the same function, aside from lengthening the code base, may also
create issues if needed to be modified, potentially creating different instances of functions
that were supposed to do the same task. Furthermore, the instances of
check_nonnegative_amount have the pub function modifier, but there is no need for it;

- [x] 4. In the interface module, a trait is defined for specifying the functions that should be available in all adapters available for the protocol. This trait is named SoroswapAggregatorAdapterTrait , and its nomenclature doesn't reflect exactly what its purpose is, as it may remit to an adapter to the Soroswap Aggregator itself;

- [x] 5. Aggregator external function initialize does not call check_initialized helper
function and calls is_initialized storage function directly. Not only is this resulted in two
different functions of similar purpose, but it may also be confusing for developers when
modifying one of the functions, potentially expecting (by a mixup) that the changes should be reglected on the other.

- [x] 6. The Aggregator can control the access to the Adapters by setting them as paused or not. Still, an Adapter can be directly called and used by users if called directly, instead of going through the Aggregator.

**Recommendations**
For each of the topics elaborated above, we recommend implementing the following
approaches into the protocol's contracts:

- [x]  1. Use function names that correctly reflect the purpose of the operations they perform;

- [x]  2. Validate the overflow possibility into the smart contract code;
- [x]  3. Avoid duplicates in the codebase by creating a shared module among the contracts, and if
a function isn't used outside of the module in which it is defined, do not declare it as public;
- [x]  4. We recommend changing it to a more precise and simple name such as AdapterTrait to
improve code readability;
- [x]  5. Functions should exist for exclusive purposes, so avoid having two or more different
functions with similar purposes to one another;
- [x]  6. Move the paused property to the Adapter and validate if they are paused within the
Adapter itself.

**Comment by joaquinsoza** [comment]
### 6

We appreciate the recommendation to move the paused property to the Adapter and validate it within the Adapter itself. However, we believe this change is unnecessary for the following reasons:

**1.	Purpose of the Adapter**:
	•	The Adapter’s sole purpose is to serve as an interface for the protocol. Introducing a paused state within the Adapter would add unnecessary complexity, such as the need to establish an admin role to manage this state.
**2.	Flexibility for Users**:
	•	We intend for the Adapter to remain flexible and reusable. By allowing the Adapter to be called directly, we provide other developers the freedom to utilize it independently if they choose. Imposing a paused state and related access controls would limit this flexibility.
**3.	Simplicity and Maintainability**:
	•	Our goal is to keep the Adapter simple and maintainable. Adding a paused state to the Adapter would introduce additional code and governance mechanisms, which could complicate its usage without providing significant benefits.

For these reasons, we believe that keeping the paused property within the Aggregator is sufficient and aligns with our design goals.

**Comment by joaquinsoza** [comment]
### 3

Has been resolved in Issue #47

**Comment by esteblock** [comment]
All points where fixed in https://github.com/soroswap/aggregator/pull/68

However here some comments:

- A negative `amount_in_max` value in `swap_tokens_for_exact_tokens` will be allowed, as it will fail with `ExcessiveInputAmount`
Because the Aggregator checks for 
if final_amount_in > amount_in_max {
        return Err(AggregatorError::ExcessiveInputAmount);
}

- A negative `amount_out_min` value in `swap_exact_tokens_for_tokens_negative_amount_out_min` will be allowed. As it is similar of calling it with `amount_out_min=0`.
This is because then, the Aggregator checks
```
 if final_amount_out < amount_out_min {
     return Err(AggregatorError::InsufficientOutputAmount);
 }
```
